### ISSUE TITLE
Can we slim down "release" builds?

### ISSUE THREAD

**Issue by kylef** [apiaryio/protagonist#241]
When you `npm install protagonist` into a project, it takes up around 100MB, over 90MB of that is built artefacts which are not needed by Protagonist at runtime, all that is really needed from `build` at runtime is `build/release/protagonist.node`.

<img width="464" alt="Screenshot 2019-07-01 at 13 38 17" src="https://user-images.githubusercontent.com/44164/60437100-831d0f00-9c05-11e9-9e42-ab7c854fafaf.png">

**Comment by kylef** [comment]
Posting as a question, not sure how feasible it is. Perhaps we can hook into GYP to delete some of the build state for release builds.

**Comment by kylef** [comment]
In theory, something like the following should contain everything for the runtime: <img width="407" alt="Screenshot 2019-07-01 at 13 51 28" src="https://user-images.githubusercontent.com/44164/60437948-5964e780-9c07-11e9-980e-e81ce8bdf2fc.png">


We do need to consider that some of the LICENSE files for vendored dependencies are no there and how to ensure they are always present.

**Comment by kylef** [comment]
Looking at the GYP documentation, doesn't appear to be any hooks or way we can execute code, and even if we could it would be tricky to make it platforn independent. I think an NPM hook for `postinstall` may be the best option, we can produce a Node file which contains instructions to delete the artefacts for Release builds (we can omit these for debug).

**Comment by pksunkara** [comment]
I thought we need `*.a` files because protagonist.node is dynamically linked. Did you get it working after deleting those things after installing protagonist as a module?

**Comment by kylef** [comment]
Yes, it works. `.a` is a static library, `protagonist.node` does not dynamically link to anything other than libc++ and system libraries:

```
$ otool -L build/Release/protagonist.node
build/Release/protagonist.node:
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 400.9.4)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.250.1)
```

**Comment by gap777** [comment]
Any progress on this? We'd love to take advantage of this slimmed down release.

**Comment by kylef** [comment]
@gap777 There hasn't been, I am not entirely sure if it is possible from Protagonist alone. I haven't had chance to fully look into it, it might be up for discussion upstream in Node JS (node-gyp specifically: https://github.com/nodejs/node-gyp) since this can apply to all the C++ add-ons for Node.

If you want a slimmed down build, you can take a look at drafter.js, it's a pure JavaScript version of Protagonist (https://github.com/apiaryio/drafter.js/).
